File: backend-java-backup/src/main/java/com/cyber/portal/controller/UserController.java
package com.cyber.portal.controller;

import com.cyber.portal.dto.PasswordChangeRequest;
import com.cyber.portal.dto.ProfileUpdateRequest;
import com.cyber.portal.entity.Complaint;
import com.cyber.portal.entity.ComplaintStatus;
import com.cyber.portal.entity.User;
import com.cyber.portal.service.ComplaintService;
import com.cyber.portal.service.UserService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/users")
public class UserController {

    @Autowired
    private UserService userService;

    @Autowired
    private ComplaintService complaintService;

    @GetMapping("/me")
    public ResponseEntity<?> getProfile(Authentication authentication) {
        if (authentication == null) {
            return ResponseEntity.status(401).body(Map.of("error", "Unauthorized"));
        }
        String email = authentication.getName();
        User user = userService.findByEmail(email);

        // Fetch citizen-specific complaint counts
        List<Complaint> complaints = complaintService.getCitizenComplaints(email);
        long totalComplaints = complaints.size();
        long resolved = complaints.stream().filter(c -> c.getStatus() == ComplaintStatus.RESOLVED).count();
        long pending = complaints.stream().filter(c -> c.getStatus() != ComplaintStatus.RESOLVED && c.getStatus() != ComplaintStatus.REJECTED).count();

        Map<String, Object> profileResponse = new HashMap<>();
        profileResponse.put("id", user.getId());
        profileResponse.put("name", user.getName());
        profileResponse.put("email", user.getEmail());
        profileResponse.put("phone", user.getPhone());
        profileResponse.put("role", user.getRole().name());
        profileResponse.put("createdAt", user.getCreatedAt());
        profileResponse.put("totalComplaints", totalComplaints);
        profileResponse.put("resolvedComplaints", resolved);
        profileResponse.put("pendingComplaints", pending);

        return ResponseEntity.ok(profileResponse);
    }

    @PutMapping("/me")
    public ResponseEntity<?> updateProfile(Authentication authentication, @Valid @RequestBody ProfileUpdateRequest request) {
        if (authentication == null) {
            return ResponseEntity.status(401).body(Map.of("error", "Unauthorized"));
        }
        String email = authentication.getName();
        User updated = userService.updateProfile(email, request);
        return ResponseEntity.ok(updated);
    }

    @PutMapping("/change-password")
    public ResponseEntity<?> changePassword(Authentication authentication, @Valid @RequestBody PasswordChangeRequest request) {
        if (authentication == null) {
            return ResponseEntity.status(401).body(Map.of("error", "Unauthorized"));
        }
        String email = authentication.getName();
        userService.changePassword(email, request);
        return ResponseEntity.ok(Map.of("message", "Password changed successfully"));
    }
}

File: backend/src/controllers/userController.js
const bcrypt = require('bcryptjs');
const { User, Complaint, AuditLog } = require('../models');

async function getProfile(req, res) {
  try {
    const user = req.user;
    
    // Fetch citizen's complaints count
    const complaints = await Complaint.findAll({ where: { userId: user.id } });
    const totalComplaints = complaints.length;
    const resolvedComplaints = complaints.filter(c => c.status === 'RESOLVED').length;
    const pendingComplaints = complaints.filter(c => c.status !== 'RESOLVED' && c.status !== 'REJECTED').length;

    return res.status(200).json({
      id: user.id,
      name: user.name,
      email: user.email,
      phone: user.phone,
      role: user.role,
      createdAt: user.createdAt,
      totalComplaints,
      resolvedComplaints,
      pendingComplaints,
    });
  } catch (error) {
    console.error('Get profile error:', error);
    return res.status(500).json({ error: 'Internal Server Error' });
  }
}

async function updateProfile(req, res) {
  try {
    const user = req.user;
    const { name, phone } = req.body;

    if (!name || !phone) {
      return res.status(400).json({ error: 'Name and phone are required' });
    }

    user.name = name;
    user.phone = phone;
    await user.save();

    // Log update profile
    try {
      await AuditLog.create({
        userId: user.id,
        userName: user.name,
        action: 'PROFILE_UPDATED',
        details: `Updated account details: name=${user.name}, phone=${user.phone}`,
        ipAddress: req.ip || req.headers['x-forwarded-for'] || '127.0.0.1'
      });
    } catch (err) {
      console.error('Failed to log profile update audit:', err);
    }


    return res.status(200).json({
      id: user.id,
      name: user.name,
      email: user.email,
      phone: user.phone,
      role: user.role,
      createdAt: user.createdAt,
    });
  } catch (error) {
    console.error('Update profile error:', error);
    return res.status(500).json({ error: 'Internal Server Error' });
  }
}

async function changePassword(req, res) {
  try {
    const user = req.user;
    const { currentPassword, newPassword } = req.body;

    if (!currentPassword || !newPassword) {
      return res.status(400).json({ error: 'Current password and new password are required' });
    }

    const isMatch = await bcrypt.compare(currentPassword, user.password);
    if (!isMatch) {
      return res.status(400).json({ error: 'Incorrect current password' });
    }

    user.password = await bcrypt.hash(newPassword, 10);
    await user.save();

    // Log change password
    try {
      await AuditLog.create({
        userId: user.id,
        userName: user.name,
        action: 'PASSWORD_CHANGED',
        details: `Successfully changed credential passwords`,
        ipAddress: req.ip || req.headers['x-forwarded-for'] || '127.0.0.1'
      });
    } catch (err) {
      console.error('Failed to log password change audit:', err);
    }


    return res.status(200).json({ message: 'Password changed successfully' });
  } catch (error) {
    console.error('Change password error:', error);
    return res.status(500).json({ error: 'Internal Server Error' });
  }
}

module.exports = {
  getProfile,
  updateProfile,
  changePassword,
};
